tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+this [blog post by Chris Done](http://chrisdone.com/posts/data-typeable) if
+you're interested.
+
+Step 2 is not so simple as it sounds though, at least not until we examine
+the details of this business more closely.
+
+## Combinators that represent shape of data type
+
+What could they look like? Well, if I showed you the data types as they are
+right now, you would probably run away in fear, cursing the tutorial and
+Haskell's generics. I have a better idea. Let's start with the simplest
+thing possible and then iterate asking ourselves how to work around some
+more interesting use cases that we would have to support. This will force us
+to make the data types more hairy but also more powerful step by step, till
+we arrive at the definitions that are actually used.
+
+For better or worse, algebraic data types lock us into a view on the world
+that consists of sums and products (another Haskell's feature that makes
+existence of generics framework possible, generics do not work with GADTs
+for example). So we need to be able to represent the following:
+
+* Data types without constructors at all: uninhabited types like `Void`.
+  This can be described as `data V1` (note: it has no constructors).
+
+* Constructors without arguments. Let this be something like this: `data U1
+  = U1`.
+
+* Sums: `data (f :+: g) = L1 f | R1 g`.
+
+* Products: `data (f :*: g) = f :*: g`.
+
+Not bad for a start. Let's try use this representation to derive a `Functor`
+instance. Deriving a such an instance means that we should provide `fmap`
+function that looks like this:
+
+```haskell
+fmap :: (a -> b) -> Rep f -> Rep f
+```
+
+…where `Rep f` maps to the type of generic representation of `f :: * -> *`.
+There is a problem though. You see, the functors' inner type that goes from
+`a` to `b` in this example has no place to live in the representation! Our
+current approach will work only for type classes that work with things of
+kind `*`, such as `Show`, but if we want to use this system to derive a
+`Functor` instance, we need to allow it to work with kind `* -> *`.
+
+We want to be able to derive `Functor` (and similar) instances generically,
+so we have to fix things up. The solution is to add one more type parameter
+`p` to all our combinators:
+
+```haskell
+data V1 p
+data U1 p = U1
+data (f :+: g) p = L1 (f p) | R1 (g p)
+data (f :*: g) p = (f p) :*: (g p)
+```
+
+This way `fmap` would be:
+
+```haskell
+fmap :: (a -> b) -> Rep f a -> Rep f b
+--                  p  =  a    p  =  b
+```
+
+But what happens to the type classes that work with `*` kinds? Here are our
+choices:
+
+* Have a separate set of combinator types for each case (`*` and `* -> *`
+  kinds).
+
+* Use the most general form (with `p`), but for `*` kinds just treat the
+  extra `p` parameter as a dummy type index that has no meaning.
+
+The authors of the generic extension went with the second option, and I
+can't blame them. We'll see that there are a lot of wrappers already and we
```

Comment:
Yes, this is a bit clearer on a second look.  The first option would define a whole other set of "wrappers" on top of the ones we have already defined.